lukefi/metsi/forestry/r/lm_tree_generation.R

```diff
@@ -94,39 +121,121 @@ Hpred<-function(lpuut,kpuut,malli,m) {
         D<-getVarCov(malli)
         sigma<-malli$sigma
         delta<-attr(malli$apVar,"Pars")[4]
-        varFix<-sigma^2*log(lpuut$HGM/lpuut$DGM+2)^(2*delta)+diag(ZLp%*%D%*%t(ZLp))
+        if (sum(names(lpuut)=="DDY")) lpuut$invDDY<-1000/lpuut$DDY
+        if (sum(names(lpuut)=="HGM")) lpuut$lnHg_Dg_2<-log(lpuut$HGM/lpuut$DGM+2)
+        varFix<-sigma^2*model.matrix(as.formula(malli$call[[5]][[2]]),data=lpuut)[,-1]^(2*delta)+diag(ZLp%*%D%*%t(ZLp))
         predHFix<-lpuut$lpm^m/muLp^m+1.3
         predHFixBC<-predHFix+0.5*m*(m+1)*lpuut$lpm^m/muLp^(m+2)*varFix
         lpuut$predHFix<-predHFix
         lpuut$predHFixBC<-lpuut$pred<-predHFixBC
+#        ok<-TRUE
         if (nrow(kpuut)>0) {
+           if (sum(names(kpuut)=="DDY")) kpuut$invDDY<-1000/kpuut$DDY
+           if (sum(names(kpuut)=="HGM")) kpuut$lnHg_Dg_2<-log(kpuut$HGM/kpuut$DGM+2)
            kpuut$Y<-1
            mu<-predict(malli,newdata=kpuut,level=0)
            y<-kpuut$lpm/(kpuut$height-1.3)^(1/m)
            Z<-model.matrix(malli$modelStruct$reStruct,data=kpuut)
-           R<-sigma^2*log(kpuut$HGM/kpuut$DGM+2)^(2*delta)
+           R<-sigma^2*model.matrix(as.formula(malli$call[[5]][[2]]),data=kpuut)[,-1]^(2*delta)
            if (length(R)>1) R<-diag(R)
            b<-D%*%t(Z)%*%solve(Z%*%D%*%t(Z)+R)%*%(y-mu)
            varb<-D-D%*%t(Z)%*%solve(Z%*%D%*%t(Z)+R)%*%Z%*%D
-           # ennustet lukupuille     
+           # ennustet lukupuille
+           alpha<-1
            muPlusbLp<-muLp+ZLp%*%b
-           varFixRan<-sigma^2*log(lpuut$HGM/lpuut$DGM+2)^(2*delta)+diag(ZLp%*%varb%*%t(ZLp))
+           # Tämä silmukka varmistaa, että kalibrointi ei väännä pituuskäyrää laskevaksi.
+           # Edellyttää että lukupuut on jrjestetty nousevasti. 
+#           while (!all(lpuut$lpm/muPlusbLp==cummax(lpuut$lpm/muPlusbLp))) {
+#                 alpha<-alpha*0.9
+#                 muPlusbLp<-muLp+alpha*ZLp%*%b
+#                 ok<-FALSE
+#                 } 
```

Comment:
Is this comment necessary?

Reply:
In general the comment lines could be checked and unnecessary should be deleted.